File: backend/app/services/access_request_service.py
from sqlalchemy.orm import Session
from sqlalchemy import and_
from typing import List, Optional
from app.models.access_request import AccessRequest
from app.models.usuario import Usuario
from app.schemas.access_request import AccessRequestCreate, AccessRequestUpdate
import uuid


class AccessRequestService:

    @staticmethod
    def create_request(db: Session, request_data: AccessRequestCreate, user_id: str) -> AccessRequest:
        """Criar uma nova requisição de acesso"""
        # Verificar se já existe uma requisição pendente para este usuário
        existing_request = db.query(AccessRequest).filter(
            and_(
                AccessRequest.user_id == user_id,
                AccessRequest.status == "PENDENTE"
            )
        ).first()

        if existing_request:
            raise ValueError("Usuário já possui uma requisição de acesso pendente")

        access_request = AccessRequest(
            user_id=user_id,
            nivel_solicitado=request_data.nivel_solicitado.value,
            trabalha_cglic=request_data.trabalha_cglic,
            justificativa=request_data.justificativa,
            status="PENDENTE"
        )

        db.add(access_request)
        db.commit()
        db.refresh(access_request)
        return access_request

    @staticmethod
    def get_user_requests(db: Session, user_id: str) -> List[AccessRequest]:
        """Buscar todas as requisições de um usuário"""
        return db.query(AccessRequest).filter(AccessRequest.user_id == user_id).all()

    @staticmethod
    def get_pending_requests(db: Session, skip: int = 0, limit: int = 100) -> List[AccessRequest]:
        """Buscar todas as requisições pendentes"""
        return db.query(AccessRequest).filter(
            AccessRequest.status == "PENDENTE"
        ).offset(skip).limit(limit).all()

    @staticmethod
    def get_all_requests(db: Session, skip: int = 0, limit: int = 100) -> List[AccessRequest]:
        """Buscar todas as requisições com informações dos usuários"""
        return db.query(AccessRequest).offset(skip).limit(limit).all()

    @staticmethod
    def get_request_by_id(db: Session, request_id: str) -> Optional[AccessRequest]:
        """Buscar uma requisição por ID"""
        return db.query(AccessRequest).filter(AccessRequest.id == request_id).first()

    @staticmethod
    def approve_request(db: Session, request_id: str, admin_id: str, observacoes: Optional[str] = None) -> AccessRequest:
        """Aprovar uma requisição de acesso"""
        access_request = AccessRequestService.get_request_by_id(db, request_id)
        if not access_request:
            raise ValueError("Requisição não encontrada")

        if access_request.status != "PENDENTE":
            raise ValueError("Apenas requisições pendentes podem ser aprovadas")

        # Atualizar a requisição
        access_request.status = "APROVADA"
        access_request.aprovado_por_id = admin_id
        access_request.observacoes_admin = observacoes

        # Atualizar o nível de acesso do usuário
        user = db.query(Usuario).filter(Usuario.id == access_request.user_id).first()
        if user:
            user.nivel_acesso = access_request.nivel_solicitado

        db.commit()
        db.refresh(access_request)
        return access_request

    @staticmethod
    def reject_request(db: Session, request_id: str, admin_id: str, observacoes: Optional[str] = None) -> AccessRequest:
        """Rejeitar uma requisição de acesso"""
        access_request = AccessRequestService.get_request_by_id(db, request_id)
        if not access_request:
            raise ValueError("Requisição não encontrada")

        if access_request.status != "PENDENTE":
            raise ValueError("Apenas requisições pendentes podem ser rejeitadas")

        # Atualizar a requisição
        access_request.status = "REJEITADA"
        access_request.aprovado_por_id = admin_id
        access_request.observacoes_admin = observacoes

        db.commit()
        db.refresh(access_request)
        return access_request

    @staticmethod
    def delete_request(db: Session, request_id: str) -> bool:
        """Deletar uma requisição de acesso"""
        access_request = AccessRequestService.get_request_by_id(db, request_id)
        if not access_request:
            return False

        db.delete(access_request)
        db.commit()
        return True

File: frontend/src/services/access-request.service.ts
import api from './api';

export interface AccessRequest {
  id: string;
  user_id: string;
  nivel_solicitado: string;
  trabalha_cglic: boolean;
  justificativa?: string;
  status: 'PENDENTE' | 'APROVADA' | 'REJEITADA';
  observacoes_admin?: string;
  aprovado_por_id?: string;
  created_at: string;
  updated_at?: string;
  user_nome?: string;
  user_email?: string;
  aprovado_por_nome?: string;
}

export interface AccessRequestCreate {
  nivel_solicitado: string;
  trabalha_cglic: boolean;
  justificativa?: string;
}

export interface AccessRequestUpdate {
  status: string;
  observacoes_admin?: string;
}

export interface AccessRequestListItem {
  id: string;
  user_nome: string;
  user_email: string;
  nivel_solicitado: string;
  trabalha_cglic: boolean;
  status: string;
  created_at: string;
}

class AccessRequestService {
  private baseURL = '/api/v1/access-requests/';

  async createRequest(data: AccessRequestCreate): Promise<AccessRequest> {
    const response = await api.post(this.baseURL, data);
    return response.data;
  }

  async getMyRequests(): Promise<AccessRequest[]> {
    const response = await api.get(`${this.baseURL}my-requests`);
    return response.data;
  }

  async getPendingRequests(): Promise<AccessRequestListItem[]> {
    const response = await api.get(`${this.baseURL}pending`);
    return response.data;
  }

  async getAllRequests(skip: number = 0, limit: number = 100): Promise<AccessRequestListItem[]> {
    const response = await api.get(`${this.baseURL}?skip=${skip}&limit=${limit}`);
    return response.data;
  }

  async getRequestDetails(requestId: string): Promise<AccessRequest> {
    const response = await api.get(`${this.baseURL}${requestId}`);
    return response.data;
  }

  async approveRequest(requestId: string, observacoes?: string): Promise<AccessRequest> {
    const response = await api.post(`${this.baseURL}${requestId}/approve`, {
      status: 'APROVADA',
      observacoes_admin: observacoes,
    });
    return response.data;
  }

  async rejectRequest(requestId: string, observacoes?: string): Promise<AccessRequest> {
    const response = await api.post(`${this.baseURL}${requestId}/reject`, {
      status: 'REJEITADA',
      observacoes_admin: observacoes,
    });
    return response.data;
  }

  async deleteRequest(requestId: string): Promise<void> {
    await api.delete(`${this.baseURL}${requestId}`);
  }
}

export const accessRequestService = new AccessRequestService();